src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp

```diff
@@ -498,6 +501,262 @@ void runtime::gpu::CudaKernelBuilder::get_softmax_op(codegen::CodeWriter& writer
+    };
+
+    auto max_lambda = [&]() { writer << "r_max = r_max > input_i ? r_max : input_i;\n"; };
+
+    auto divide_lambda = [&]() {
+        writer << "input_i = expf(input_i - r_max) / r_sum;\n";
+        writer << "out[input_idx] = input_i;\n";
+    };
+
+    writer << "extern \"C\" __global__ void cuda_" << name << args.get_input_signature();
+    writer.block_begin();
+    {
+        writer << "extern __shared__ " << data_types[1] << " sdata[];\n";
+        if (non_reduce_rank > 0)
+        {
+            writer << "uint32_t bid = blockIdx.x;\n";
+        }
+        writer << "uint32_t tid = threadIdx.x;\n";
+        writer << "uint32_t step = blockDim.x; \n";
+        collective_coordinate_transform_helper(writer,
+                                               "bid",
+                                               "non_reduce_strides",
+                                               "non_reduce_strides_magic",
+                                               "non_reduce_strides_shift",
+                                               "non_reduce_strides_in_input",
+                                               "non_reduce_coordinate",
+                                               non_reduce_rank,
+                                               true,
+                                               "non_reduce_input_index");
+        writer << "uint32_t input_idx;\n";
+        writer << "uint32_t reduce_idx = tid;\n";
+        writer << data_types[1] << " r_max;\n";
+        writer << data_types[1] << " input_i;\n";
+
+        // find max
+        writer.block_begin();
+        {
+            get_reduce_input_lambda();
+            writer << "r_max = input_i;\n";
+            writer << "reduce_idx += step;\n";
+        }
+        writer.block_end();
+        writer << "while (reduce_idx + 7 * step < reduce_count)\n";
+        writer.block_begin();
+        {
+            for (int i = 0; i < 8; i++)
+            {
+                writer.block_begin();
+                get_reduce_input_lambda();
+                max_lambda();
+                writer << "reduce_idx += step;\n";
+                writer.block_end();
+            }
+        }
+        writer.block_end();
+
+        writer << "while (reduce_idx < reduce_count)\n";
+        writer.block_begin();
+        {
+            writer.block_begin();
+            get_reduce_input_lambda();
+            max_lambda();
+            writer << "reduce_idx += step;\n";
+            writer.block_end();
+        }
+        writer.block_end();
+        // reduction max
+        // accumulate 32 threads for each warp
+        for (int i = 16; i >= 1; i >>= 1)
+        {
+            if (block_size_x > i)
+            {
+                writer << "input_i = __shfl_down_sync(0xffffffff, r_max, " << i << ", 32);\n";
+                max_lambda();
+            }
+        }
+        if (block_size_x > 32)
+        {
+            writer << "uint32_t lane_idx = threadIdx.x & 0x1f; \n";
```

Comment:
nit: How about 31, or even easier to read `#define MAX_LANE_IDX 31` and 
```
            writer << "uint32_t lane_idx = threadIdx.x & MAX_LANE_IDX; \n";
```

Reply:
good point.

Reply:
Can we also add
#define MAX_LANE_IDX 31 here: https://github.com/NervanaSystems/ngraph/blob/f33317cc5538ddd7ff9e3174429846e80828b01c/src/ngraph/runtime/gpu/nvrtc/helpers.hpp#L58
And the replace the 0x1f above with it?